codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -0,0 +1,319 @@
+                    @Override
+                    public void run() {
+                        childOnHeadersRead(streamInfo, headers, endOfStream);
+                    }
+                });
+            }
+        }
+
+        void childOnHeadersRead(StreamInfo streamInfo, Http2Headers headers, boolean endOfStream) {
+            fireChildReadAndRegister(streamInfo, headers);
+            if (endOfStream) {
+                // TODO
+            }
+        }
+
+        @Override
+        public int onDataRead(ChannelHandlerContext ctx, int streamId, final ByteBuf data,
+                int padding, final boolean endOfStream) throws Http2Exception {
+            Http2Stream stream = http2Handler.connection().requireStream(streamId);
+            final StreamInfo streamInfo = (StreamInfo) stream.getProperty(StreamInfo.class);
+            EventLoop childLoop = streamInfo.childChannel.eventLoop();
+            if (childLoop.inEventLoop()) {
+                childOnDataRead(streamInfo, data, endOfStream);
+            } else {
+                childLoop.execute(new Runnable() {
+                    @Override
+                    public void run() {
+                        childOnDataRead(streamInfo, data, endOfStream);
+                    }
+                });
+            }
+            return data.readableBytes() + padding;
+        }
+
+        void childOnDataRead(StreamInfo streamInfo, ByteBuf data, boolean endOfStream) {
+            fireChildReadAndRegister(streamInfo, data);
+            if (endOfStream) {
+                // TODO
+            }
+        }
+    }
+
+    static class StreamInfo {
+        final Http2StreamChannel childChannel;
+        /**
+         * True if stream is in {@link Http2MultiplexCodec#steamsToFireChildReadComplete}.
+         */
+        boolean inStreamsToFireChildReadComplete;
+
+        StreamInfo(Http2StreamChannel childChannel) {
+            this.childChannel = childChannel;
+        }
+    }
+
+    class Http2StreamChannel extends ChildChannel {
+        private final Http2Stream stream;
+
+        Http2StreamChannel(Http2Stream stream) {
+            super(Http2MultiplexCodec.this.ctx.channel());
+            this.stream = stream;
+        }
+
+        @Override
+        public void doClose() throws Exception {
+            EventLoop parentLoop = ctx.channel().eventLoop();
+            if (parentLoop.inEventLoop()) {
+                parentClose();
+            } else {
+                parentLoop.execute(new Runnable() {
+                    @Override
+                    public void run() {
+                        parentClose();
+                    }
+                });
+            }
+            super.doClose();
+        }
+
+        public void parentClose() {
+            try {
+                if (stream.state() != Http2Stream.State.CLOSED) {
+                    http2Handler.resetStream(ctx, stream.id(), Http2Error.CANCEL.code(),
+                        ctx.newPromise());
```

Comment:
The promise handling will need some thought behind them. For a while I considered just passing child promises to the parent, but that seems like it breaks some (maybe broken) assumptions as to what sort of promises handlers may expect to receive. We could create new promises for the parent channel and complete the child promises when the parent promises complete.

But doWrite has the more painful of the Promise issues. ChannelOutboundBuffer does not let you get access to the ChannelPromises that it contains and will make sure to recycle itself, so you can't "steal" the ChannelPromises and you are expected to process them immediately. To deal with this we could override AbstractChannel's buffering logic or we could "punch a whole" in the ChannelOutboundBuffer API to get us access to the promises. Although I'm mostly leaning toward adding a way to "transfer" between two ChannelOutboundBuffers. It would still be a copy, but the main alternative involves synchronization.